src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs

```diff
@@ -53,13 +54,15 @@ public class DefaultHtmlGenerator : IHtmlGenerator
         /// <param name="urlHelperFactory">The <see cref="IUrlHelperFactory"/>.</param>
         /// <param name="htmlEncoder">The <see cref="HtmlEncoder"/>.</param>
         /// <param name="validationAttributeProvider">The <see cref="ValidationHtmlAttributeProvider"/>.</param>
+        /// <param name="validationCssClassNameProvider">The <see cref="IValidationCssClassNameProvider"/>.</param>
         public DefaultHtmlGenerator(
             IAntiforgery antiforgery,
             IOptions<MvcViewOptions> optionsAccessor,
             IModelMetadataProvider metadataProvider,
             IUrlHelperFactory urlHelperFactory,
             HtmlEncoder htmlEncoder,
-            ValidationHtmlAttributeProvider validationAttributeProvider)
+            ValidationHtmlAttributeProvider validationAttributeProvider,
+            IValidationCssClassNameProvider validationCssClassNameProvider)
```

Comment:
Could we make these properties on `HtmlHelperOptions` rather than a separate type? We generally don't expect users to customize options by injecting new DI services. That also avoids a bunch of source / binary breaking changes that were introduced as part of this PR.

Reply:
Ah, yes, I had not spotted that type. That sounds like a good alternative that has less impact on the code base.

What binary breaking change do you refer to? Because the HtmlHelper properties will still be removed, right?

Reply:
We would keep the HtmlHelper properties as is since removing them would be breaking. Adding a new constructor parameter is also a breaking change.

Reply:
I agree this is a much better approach. I implemented it. I did not yet re-add the static fields to HtmlHelper, as it might be confusing for people that these variables are defined there while they are not used anymore... Should we annotate them Obsolete, or add some comment?